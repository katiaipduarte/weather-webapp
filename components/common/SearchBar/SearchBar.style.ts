import styled from 'styled-components'

export const SearchBarContainer = styled.form`
    max-width: 24rem;
    padding: 0.5rem 0;
    margin: 1rem 0;
    border-bottom: 1px solid white;
    margin-bottom: 2.25rem;

    &:focus-visible,
    &:focus:not(:focus-visible) {
      outline: auto;
      color: white;
    }

    #search-term-btn {
      position: absolute;
      top: 0;
      right: 0;
      width: 4rem;
      height: 4rem;
      filter: brightness(75%);
      cursor: pointer;
      border: none;
      display: none;
    }

    #clear-search-term-btn {
      background: none;
      border: none;
      cursor: pointer;
      vertical-align: middle;
      svg {
        color: white;
        height: 1rem;
      }
      &:focus {
        outline: none;
      }
    }

    input {
      border: none;
      color: white;
      outline: none;
      font-size: 1rem;
      width: calc(100% - 1rem);
      background: transparent;
      ::placeholder {
        color: white;
      }
    }
  }

  @media (min-width: 700px) {
    #search-term-btn {
      display: inherit;
    }
  }
`
