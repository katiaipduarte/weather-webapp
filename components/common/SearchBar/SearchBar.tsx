import LocationList from '@components/common/LocationList/LocationList'
import { faSearch, faTimes } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { GPSLocation } from '@interfaces/open-weather-api/location'
import { searchForLocations } from '@services/search-locations.service'
import lodash from 'lodash'
import { ChangeEvent, FormEvent, useCallback, useState } from 'react'
import { SearchBarContainer } from './SearchBar.style'

const SearchBar = (): JSX.Element => {
	const [searchTerm, setSearchTerm] = useState<string>('')
	const [showClearButton, setShowClearButton] = useState<boolean>(false)
	const [locations, setLocations] = useState<GPSLocation[]>([])
	const [loading, setLoading] = useState<boolean>(true)
	const [showOptions, setShowOptions] = useState<boolean>(false)

	const handleClearSearch = (): void => {
		setSearchTerm('')
		setShowClearButton(false)
		setShowOptions(false)
	}

	const handleChange = async (
		event: ChangeEvent<HTMLInputElement>
	): Promise<void> => {
		setShowClearButton(event.target.value.length !== 0)
		setSearchTerm(event.target.value)

		if (event.target.value.length >= 3) {
			setLoading(true)
			debounce(event.target.value)
		}

		setShowOptions(event.target.value.length >= 3)
	}

	const handleSubmit = (event: FormEvent<HTMLFormElement>): void => {
		event.preventDefault()

		setSearchTerm(searchTerm)
	}

	const debounce = useCallback(
		lodash.debounce(async (searchTerm: string) => {
			const res = await searchForLocations(searchTerm)
			const list: GPSLocation[] = res ? (res as GPSLocation[]) : []
			setLocations(list)
			setLoading(false)
		}, 500),
		[]
	)

	return (
		<>
			<SearchBarContainer
				tabIndex={-1}
				role='searchbox'
				aria-label='Search location form'
				onSubmit={(e) => handleSubmit(e)}
			>
				<button
					id='search-term-btn'
					aria-label='Click to search'
					tabIndex={-1}
					type='submit'
				>
					<FontAwesomeIcon icon={faSearch} />
				</button>

				<input
					type='text'
					placeholder='Another location'
					value={searchTerm}
					maxLength={50}
					onChange={handleChange}
					role='search'
					aria-label='Another location'
				/>
				<button
					id='clear-search-term-btn'
					aria-label='Click to clear search'
					onClick={handleClearSearch}
					style={{ visibility: showClearButton ? 'initial' : 'hidden' }}
					tabIndex={0}
				>
					<FontAwesomeIcon icon={faTimes} />
				</button>
			</SearchBarContainer>
			{!loading && showOptions && <LocationList locations={locations} />}
		</>
	)
}

export default SearchBar
