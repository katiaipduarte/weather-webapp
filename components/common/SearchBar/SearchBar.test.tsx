import { fireEvent, render, screen } from '@testing-library/react'
import 'jest-styled-components'
import SearchBar from './SearchBar'

describe('<SearchBar /> component Unit Test', () => {
	it('should render Searchbar', () => {
		const { container } = render(<SearchBar />)
		expect(container).toMatchSnapshot()
	})

	it('should render 1 <SearchBar /> component', () => {
		render(<SearchBar />)
		expect(screen.getByRole('searchbox')).toBeInTheDocument()
	})

	it('should not show clear button if input is empty', () => {
		render(<SearchBar />)
		const searchInput = screen.getByPlaceholderText(
			'Another location'
		) as HTMLInputElement
		expect(searchInput.value).toBe('')
		expect(
			screen.getByLabelText('Click to clear search').style.visibility
		).toBe('hidden')
	})

	it('should show clear button if input is not empty', () => {
		render(<SearchBar />)
		const searchInput = screen.queryByPlaceholderText(
			'Another location'
		) as HTMLInputElement
		fireEvent.change(searchInput, { target: { value: 'test' } })
		expect(searchInput.value).toBe('test')
		expect(
			screen.getByLabelText('Click to clear search').style.visibility
		).toBe('initial')
	})

	it('should hide the clear button on click', () => {
		render(<SearchBar />)
		const clearBtn = screen.getByLabelText('Click to clear search')
		expect(clearBtn.style.visibility).toBe('hidden')

		const searchInput = screen.queryByPlaceholderText(
			'Search comics by character'
		) as HTMLInputElement
		fireEvent.change(searchInput, { target: { value: 'test' } })
		expect(clearBtn.style.visibility).toBe('initial')

		fireEvent.click(clearBtn)

		expect(clearBtn.style.visibility).toBe('hidden')
	})

	it('should render an input of type text and allows at max 50 char', () => {
		render(<SearchBar />)
		const searchInput = screen.queryByPlaceholderText(
			'Search comics by character'
		) as HTMLInputElement
		expect(searchInput.type).toBe('text')
		expect(searchInput.maxLength).toBe(50)

		const validInput = 'ygevhodxbvbaqzpflfkoylzhkbcerwmkavypauvvdwpeunagmq'
		fireEvent.change(searchInput, { target: { value: validInput } })

		expect(validInput).toHaveLength(50)
		expect(searchInput.value).toBe(validInput)

		fireEvent.keyDown(searchInput, { key: 'a' })
		expect(searchInput.value).toBe(validInput)
	})
})
