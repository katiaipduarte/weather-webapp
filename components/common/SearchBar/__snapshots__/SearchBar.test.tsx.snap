// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> component Unit Test should render Searchbar 1`] = `
.c0 {
  max-width: 24rem;
  padding: 0.5rem 0;
  margin: 1rem 0;
  border-bottom: 1px solid white;
  margin-bottom: 3.25rem;
}

.c0:focus-visible,
.c0:focus:not(:focus-visible) {
  outline: auto;
  color: white;
}

.c0 #search-term-btn {
  position: absolute;
  top: 0;
  right: 0;
  width: 4rem;
  height: 4rem;
  -webkit-filter: brightness(75%);
  filter: brightness(75%);
  cursor: pointer;
  border: none;
}

.c0 #clear-search-term-btn {
  background: none;
  border: none;
  cursor: pointer;
  vertical-align: middle;
}

.c0 #clear-search-term-btn svg {
  color: white;
  height: 1rem;
}

.c0 #clear-search-term-btn:focus {
  outline: none;
}

.c0 input {
  border: none;
  color: white;
  outline: none;
  font-size: 1rem;
  width: calc(100% - 1rem);
  background: transparent;
}

.c0 input::-webkit-input-placeholder {
  color: white;
}

.c0 input::-moz-placeholder {
  color: white;
}

.c0 input:-ms-input-placeholder {
  color: white;
}

.c0 input::placeholder {
  color: white;
}

<div>
  <form
    aria-label="Search location form"
    class="c0"
    role="searchbox"
    tabindex="-1"
  >
    <button
      aria-label="Click to search"
      id="search-term-btn"
      tabindex="-1"
      type="submit"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-magnifying-glass "
        data-icon="magnifying-glass"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M500.3 443.7l-119.7-119.7c27.22-40.41 40.65-90.9 33.46-144.7C401.8 87.79 326.8 13.32 235.2 1.723C99.01-15.51-15.51 99.01 1.724 235.2c11.6 91.64 86.08 166.7 177.6 178.9c53.8 7.189 104.3-6.236 144.7-33.46l119.7 119.7c15.62 15.62 40.95 15.62 56.57 0C515.9 484.7 515.9 459.3 500.3 443.7zM79.1 208c0-70.58 57.42-128 128-128s128 57.42 128 128c0 70.58-57.42 128-128 128S79.1 278.6 79.1 208z"
          fill="currentColor"
        />
      </svg>
    </button>
    <input
      aria-label="Another location"
      maxlength="50"
      placeholder="Another location"
      role="search"
      type="text"
      value=""
    />
    <button
      aria-label="Click to clear search"
      id="clear-search-term-btn"
      style="visibility: hidden;"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-xmark "
        data-icon="xmark"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M310.6 361.4c12.5 12.5 12.5 32.75 0 45.25C304.4 412.9 296.2 416 288 416s-16.38-3.125-22.62-9.375L160 301.3L54.63 406.6C48.38 412.9 40.19 416 32 416S15.63 412.9 9.375 406.6c-12.5-12.5-12.5-32.75 0-45.25l105.4-105.4L9.375 150.6c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0L160 210.8l105.4-105.4c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25l-105.4 105.4L310.6 361.4z"
          fill="currentColor"
        />
      </svg>
    </button>
  </form>
</div>
`;
